connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java

```diff
@@ -307,7 +307,7 @@ void doShutdown() {
                                     + " as the connector has been scheduled for shutdown"),
                         null);
             }
-            if (state == State.STARTED)
+            if (state == State.STARTED || state == State.FAILED)
                 connector.stop();
```

Comment:
connector.stop can fail, and if that happens then we don't transit to STOPPED, we re-transit to FAILED, but have a different error. I think this will shadow the exception that caused the connector to fail in the first place, which is almost certainly more confusing than helpful.

In the catch clause, can you swap out the state/statusListener for onFailure? It has failure-deduplication logic that seems useful here.

Reply:
I think that some of the tests in WorkerConnectorTest exercising failed connectors will need some verifyCleanShutdown assertions updated. A test which tests exceptions in both start() and stop() would also be very helpful :)